src/gba/vfame.c

```diff
@@ -0,0 +1,324 @@
+				cart->acceptingModeChange = false;
+			}
+		}
+	}
+
+	// If we are in "mode change mode" we can change either SRAM or ROM modes
+	// Currently unknown if other SRAM writes in this mode should have any effect
+	if (cart->acceptingModeChange) {
+		if (address == 0x0E00FFFE) {
+			cart->sramMode = value;
+		} else if (address == 0xE00FFFD) {
+			cart->romMode = value;
+		}
+	}
+
+	if (cart->sramMode == -1) {
+		// when SRAM mode is uninitialised you can't write to it
+		return;
+	}
+
+	// if mode has been set - the address and value of the SRAM write will be modified
+	address = _modifySramAddress(cart->cartType, address, cart->sramMode);
+	value = _modifySramValue(cart->cartType, value, cart->sramMode);
+	// these writes are mirrored
+	if (address >= 0xE008000) {
+		address -= 0x8000;
+	}
+	sramData[address & (SIZE_CART_SRAM - 1)] = value;
+	sramData[(address + 0x8000) & (SIZE_CART_SRAM - 1)] = value;
+}
+
+static uint32_t _modifySramAddress(enum GBAVFameCartType type, uint32_t address, int mode) {
+	switch(mode & 0xF) {
+		case 0x00: case 0x04: case 0x08: case 0x0C:
+			return address;
+		break;
+		case 0x01: case 0x05: case 0x09: case 0x0D:
+			if (type == VFAME_GEORGE) {
+				return _reorderBits(address, ADDRESS_REORDERING_GEORGE_1, 16);
+			} else {
+				return _reorderBits(address, ADDRESS_REORDERING_1, 16);
+			}
+		break;
+		case 0x02: case 0x06: case 0x0A: case 0x0e:
+			if (type == VFAME_GEORGE) {
+				return _reorderBits(address, ADDRESS_REORDERING_GEORGE_2, 16);
+			} else {
+				return _reorderBits(address, ADDRESS_REORDERING_2, 16);
+			}
+		break;
+		case 0x03: case 0x07: case 0x0b: case 0x0F:
+			if (type == VFAME_GEORGE) {
+				return _reorderBits(address, ADDRESS_REORDERING_GEORGE_3, 16);
+			} else {
+				return _reorderBits(address, ADDRESS_REORDERING_3, 16);
+			}
+		break;
+	}
+	return address;
+}
+
+static int8_t _modifySramValue(enum GBAVFameCartType type, int8_t value, int mode) {
+	switch(mode & 0xF) {
+		case 0x00: case 0x01: case 0x02: case 0x03:
+			return value;
+		break;
+		case 0x04: case 0x05: case 0x06: case 0x07:
+			if (type == VFAME_GEORGE) {
+				return _reorderBits(value,VALUE_REORDERING_GEORGE_1, 8);
+			} else {
+				return _reorderBits(value,VALUE_REORDERING_1, 8);
+			}
+		break;
+		case 0x08: case 0x09: case 0x0A: case 0x0B:
+			if (type == VFAME_GEORGE) {
+				return _reorderBits(value,VALUE_REORDERING_GEORGE_2, 8);
+			} else {
+				return _reorderBits(value,VALUE_REORDERING_2, 8);
+			}
+		break;
+		case 0x0C: case 0x0D: case 0x0E: case 0x0F:
+			if (type == VFAME_GEORGE) {
+				return _reorderBits(value,VALUE_REORDERING_GEORGE_3, 8);
+			} else {
+				return _reorderBits(value,VALUE_REORDERING_3, 8);
+			}
+		break;
+	}
+	return value;
+}
+
+// Reorder bits in a byte according to the reordering given
+static int _reorderBits(uint32_t value, const char* reordering, int reorderLength) {
+	uint32_t retval = value;
+
+	for(int x = reorderLength; x > 0; x--)
```

Comment:
Also, please put the `int x` outside of the for loop. I've had some compilers complain about this even though it should be allowed.